bungiesearch/signals.py

```diff
@@ -1,31 +1,55 @@
 from _collections import defaultdict
+
+from django.db.models import signals
+from importlib import import_module
+
 from . import Bungiesearch
 from .utils import update_index, delete_index_item
 
+def get_signal_processor():
+    signals = Bungiesearch.BUNGIE['SIGNALS']
+    if 'SIGNAL_CLASS' in signals:
+        module_str      = signals['SIGNAL_CLASS']
+        signal_module   = import_module(module_str)
+        signal_instance = signal_module()
+    else:
+        signal_instance = BungieSignalProcessor()
+
+    return signal_instance
 
 __items_to_be_indexed__ = defaultdict(list)
-def post_save_connector(sender, instance, **kwargs):
-    try:
-        Bungiesearch.get_index(sender, via_class=True)
-    except KeyError:
-        return # This model is not managed by Bungiesearch.
-
-    try:
-        buffer_size = Bungiesearch.BUNGIE['SIGNALS']['BUFFER_SIZE']
-    except KeyError:
-        buffer_size = 100
-
-    __items_to_be_indexed__[sender].append(instance)
-
-    if len(__items_to_be_indexed__[sender]) >= buffer_size:
-        update_index(__items_to_be_indexed__[sender], sender.__name__, buffer_size)
-        # Let's now empty this buffer or we'll end up reindexing every item which was previously buffered.s
-        __items_to_be_indexed__[sender] = []
-
-def pre_delete_connector(sender, instance, **kwargs):
-    try:
-        Bungiesearch.get_index(sender, via_class=True)
-    except KeyError:
-        return # This model is not managed by Bungiesearch.
-
-    delete_index_item(instance, sender.__name__)
+class BungieSignalProcessor(object):
+
+    def post_save_connector(self, sender, instance, **kwargs):
+        try:
+            Bungiesearch.get_index(sender, via_class=True)
+        except KeyError:
+            return # This model is not managed by Bungiesearch.
+
+        try:
+            buffer_size = Bungiesearch.BUNGIE['SIGNALS']['BUFFER_SIZE']
+        except KeyError:
+            buffer_size = 100
+
+        __items_to_be_indexed__[sender].append(instance)
+
+        if len(__items_to_be_indexed__[sender]) >= buffer_size:
+            update_index(__items_to_be_indexed__[sender], sender.__name__, buffer_size)
+            # Let's now empty this buffer or we'll end up reindexing every item which was previously buffered.
+            __items_to_be_indexed__[sender] = []
+    
+    def pre_delete_connector(self, sender, instance, **kwargs):
+        try:
+            Bungiesearch.get_index(sender, via_class=True)
+        except KeyError:
+            return # This model is not managed by Bungiesearch.
+
+        delete_index_item(instance, sender.__name__)
+    
+    def setup(self, model):
+        signals.post_save.connect(self.post_save_connector, sender=model)
+        signals.pre_delete.connect(self.pre_delete_connector, sender=model)
+
+    def teardown(self, model):
+        signals.post_save.disconnect(self.post_save_connector, sender=model)
```

Comment:
The `pre_delete` signal should be disconnected prior to the `post_save` so that the teardown function is a mirror of the setup function.